File: content/develop/use-cases/leaderboard/php/RedisLeaderboard.php
<?php

use Predis\Client as PredisClient;

/**
 * Redis-backed leaderboard implementation using a sorted set and per-user metadata hashes.
 */
class RedisLeaderboard
{
    private PredisClient $redis;
    private string $key;
    private int $maxEntries;

    public function __construct(
        ?PredisClient $redis = null,
        string $key = 'leaderboard:demo',
        int $maxEntries = 100
    ) {
        $this->redis = $redis ?? new PredisClient([
            'host' => '127.0.0.1',
            'port' => 6379,
        ]);
        $this->key = $key !== '' ? $key : 'leaderboard:demo';
        $this->maxEntries = $this->normalizePositiveInt($maxEntries, 'maxEntries');
    }

    public function getKey(): string
    {
        return $this->key;
    }

    public function getMaxEntries(): int
    {
        return $this->maxEntries;
    }

    public function upsertUser(string $userId, float $score, array $metadata = []): array
    {
        $payload = $this->coerceMetadata($metadata);

        $pipeline = $this->redis->pipeline();
        $pipeline->zadd($this->key, [$userId => $score]);
        if ($payload !== []) {
            $pipeline->hmset($this->metadataKey($userId), $payload);
        }
        $pipeline->execute();

        $trimmedUserIds = $this->trimToMaxEntries();
        $entry = $this->getUserEntry($userId);

        if ($entry === null) {
            return [
                'rank' => 0,
                'user_id' => $userId,
                'score' => $score,
                'metadata' => $payload,
                'trimmed_user_ids' => $trimmedUserIds,
            ];
        }

        $entry['trimmed_user_ids'] = $trimmedUserIds;
        return $entry;
    }

    public function incrementScore(string $userId, float $amount, array $metadata = []): array
    {
        $payload = $this->coerceMetadata($metadata);

        $pipeline = $this->redis->pipeline();
        $pipeline->zincrby($this->key, $amount, $userId);
        if ($payload !== []) {
            $pipeline->hmset($this->metadataKey($userId), $payload);
        }
        $result = $pipeline->execute();
        $newScore = isset($result[0]) ? (float) $result[0] : 0.0;

        $trimmedUserIds = $this->trimToMaxEntries();
        $entry = $this->getUserEntry($userId);

        if ($entry === null) {
            return [
                'rank' => 0,
                'user_id' => $userId,
                'score' => $newScore,
                'metadata' => $payload,
                'trimmed_user_ids' => $trimmedUserIds,
            ];
        }

        $entry['trimmed_user_ids'] = $trimmedUserIds;
        return $entry;
    }

    public function setMaxEntries(int $maxEntries): array
    {
        $this->maxEntries = $this->normalizePositiveInt($maxEntries, 'maxEntries');
        return $this->trimToMaxEntries();
    }

    public function getTop(int $count): array
    {
        $normalizedCount = $this->normalizePositiveInt($count, 'count');
        $entries = $this->redis->zrevrange($this->key, 0, $normalizedCount - 1, 'WITHSCORES');
        return $this->hydrateEntries($entries, 1);
    }

    public function getAroundRank(int $rank, int $count): array
    {
        $normalizedRank = $this->normalizePositiveInt($rank, 'rank');
        $normalizedCount = $this->normalizePositiveInt($count, 'count');
        $totalEntries = $this->getSize();

        if ($totalEntries === 0) {
            return [];
        }

        if ($totalEntries <= $normalizedCount) {
            return $this->listAll();
        }

        $halfWindow = intdiv($normalizedCount, 2);
        $start = max(0, $normalizedRank - 1 - $halfWindow);
        $maxStart = $totalEntries - $normalizedCount;
        if ($start > $maxStart) {
            $start = $maxStart;
        }
        $end = $start + $normalizedCount - 1;

        $entries = $this->redis->zrevrange($this->key, $start, $end, 'WITHSCORES');
        return $this->hydrateEntries($entries, $start + 1);
    }

    public function getRank(string $userId): ?int
    {
        $rank = $this->redis->zrevrank($this->key, $userId);
        if ($rank === null) {
            return null;
        }
        return (int) $rank + 1;
    }

    public function getUserMetadata(string $userId): array
    {
        return $this->redis->hgetall($this->metadataKey($userId));
    }

    public function getUserEntry(string $userId): ?array
    {
        $pipeline = $this->redis->pipeline();
        $pipeline->zscore($this->key, $userId);
        $pipeline->zrevrank($this->key, $userId);
        $pipeline->hgetall($this->metadataKey($userId));
        $result = $pipeline->execute();

        if (!isset($result[0], $result[1]) || $result[0] === null || $result[1] === null) {
            return null;
        }

        return [
            'rank' => (int) $result[1] + 1,
            'user_id' => $userId,
            'score' => (float) $result[0],
            'metadata' => $result[2] ?? [],
            'trimmed_user_ids' => [],
        ];
    }

    public function listAll(): array
    {
        $entries = $this->redis->zrevrange($this->key, 0, -1, 'WITHSCORES');
        return $this->hydrateEntries($entries, 1);
    }

    public function getSize(): int
    {
        return (int) $this->redis->zcard($this->key);
    }

    public function deleteUser(string $userId): bool
    {
        $pipeline = $this->redis->pipeline();
        $pipeline->zrem($this->key, $userId);
        $pipeline->del([$this->metadataKey($userId)]);
        $result = $pipeline->execute();
        return isset($result[0]) && (int) $result[0] === 1;
    }

    public function clear(): void
    {
        $userIds = $this->redis->zrange($this->key, 0, -1);
        $keys = [$this->key];
        foreach ($userIds as $userId) {
            $keys[] = $this->metadataKey((string) $userId);
        }
        $this->redis->del($keys);
    }

    private function metadataKey(string $userId): string
    {
        return "{$this->key}:user:{$userId}";
    }

    private function coerceMetadata(array $metadata): array
    {
        $payload = [];
        foreach ($metadata as $field => $value) {
            $payload[(string) $field] = (string) $value;
        }
        return $payload;
    }

    private function trimToMaxEntries(): array
    {
        $overflow = $this->getSize() - $this->maxEntries;
        if ($overflow <= 0) {
            return [];
        }

        $trimmedUserIds = $this->redis->zrange($this->key, 0, $overflow - 1);
        if ($trimmedUserIds === []) {
            return [];
        }

        $this->redis->zremrangebyrank($this->key, 0, $overflow - 1);

        $metadataKeys = [];
        foreach ($trimmedUserIds as $userId) {
            $metadataKeys[] = $this->metadataKey((string) $userId);
        }
        if ($metadataKeys !== []) {
            $this->redis->del($metadataKeys);
        }

        return array_map('strval', $trimmedUserIds);
    }

    private function hydrateEntries(array $entries, int $startRank): array
    {
        $hydrated = [];
        $rank = $startRank;
        foreach ($entries as $userId => $score) {
            $hydrated[] = [
                'rank' => $rank++,
                'user_id' => (string) $userId,
                'score' => (float) $score,
                'metadata' => $this->getUserMetadata((string) $userId),
                'trimmed_user_ids' => [],
            ];
        }
        return $hydrated;
    }

    private function normalizePositiveInt(int $value, string $fieldName): int
    {
        if ($value < 1) {
            throw new InvalidArgumentException("{$fieldName} must be at least 1");
        }
        return $value;
    }
}

File: content/develop/use-cases/leaderboard/dotnet/RedisLeaderboard.cs
using StackExchange.Redis;

namespace LeaderboardDemo;

public record LeaderboardEntry(
    int Rank,
    string UserId,
    double Score,
    IReadOnlyDictionary<string, string> Metadata,
    IReadOnlyList<string> TrimmedUserIds);

public class RedisLeaderboard
{
    private readonly IDatabase _db;
    private readonly string _key;
    private int _maxEntries;

    public RedisLeaderboard(IDatabase? db = null, string key = "leaderboard:demo", int maxEntries = 100)
    {
        _db = db ?? ConnectionMultiplexer.Connect("localhost:6379").GetDatabase();
        _key = string.IsNullOrWhiteSpace(key) ? "leaderboard:demo" : key;
        _maxEntries = NormalizePositiveInt(maxEntries, nameof(maxEntries));
    }

    public string Key => _key;

    public int MaxEntries => _maxEntries;

    public LeaderboardEntry UpsertUser(
        string userId,
        double score,
        IReadOnlyDictionary<string, string>? metadata = null)
    {
        var payload = CoerceMetadata(metadata);
        var batch = _db.CreateBatch();
        var scoreTask = batch.SortedSetAddAsync(_key, userId, score);
        Task metadataTask = Task.CompletedTask;
        if (payload.Count > 0)
        {
            metadataTask = batch.HashSetAsync(
                MetadataKey(userId),
                payload.Select(pair => new HashEntry(pair.Key, pair.Value)).ToArray());
        }

        batch.Execute();
        Task.WaitAll(scoreTask, metadataTask);

        var trimmedUserIds = TrimToMaxEntries();
        var entry = GetUserEntry(userId);
        return entry is null
            ? new LeaderboardEntry(0, userId, score, payload, trimmedUserIds)
            : entry with { TrimmedUserIds = trimmedUserIds };
    }

    public LeaderboardEntry IncrementScore(
        string userId,
        double amount,
        IReadOnlyDictionary<string, string>? metadata = null)
    {
        var payload = CoerceMetadata(metadata);
        var batch = _db.CreateBatch();
        var scoreTask = batch.SortedSetIncrementAsync(_key, userId, amount);
        Task metadataTask = Task.CompletedTask;
        if (payload.Count > 0)
        {
            metadataTask = batch.HashSetAsync(
                MetadataKey(userId),
                payload.Select(pair => new HashEntry(pair.Key, pair.Value)).ToArray());
        }

        batch.Execute();
        Task.WaitAll(scoreTask, metadataTask);

        var trimmedUserIds = TrimToMaxEntries();
        var entry = GetUserEntry(userId);
        return entry is null
            ? new LeaderboardEntry(0, userId, scoreTask.Result, payload, trimmedUserIds)
            : entry with { TrimmedUserIds = trimmedUserIds };
    }

    public IReadOnlyList<string> SetMaxEntries(int maxEntries)
    {
        _maxEntries = NormalizePositiveInt(maxEntries, nameof(maxEntries));
        return TrimToMaxEntries();
    }

    public IReadOnlyList<LeaderboardEntry> GetTop(int count)
    {
        var normalizedCount = NormalizePositiveInt(count, nameof(count));
        var entries = _db.SortedSetRangeByRankWithScores(
            _key,
            0,
            normalizedCount - 1,
            Order.Descending);
        return HydrateEntries(entries, 1);
    }

    public IReadOnlyList<LeaderboardEntry> GetAroundRank(int rank, int count)
    {
        var normalizedRank = NormalizePositiveInt(rank, nameof(rank));
        var normalizedCount = NormalizePositiveInt(count, nameof(count));
        var totalEntries = (int)GetSize();

        if (totalEntries == 0)
        {
            return [];
        }

        if (totalEntries <= normalizedCount)
        {
            return ListAll();
        }

        var halfWindow = normalizedCount / 2;
        var start = Math.Max(0, normalizedRank - 1 - halfWindow);
        var maxStart = totalEntries - normalizedCount;
        if (start > maxStart)
        {
            start = maxStart;
        }

        var end = start + normalizedCount - 1;
        var entries = _db.SortedSetRangeByRankWithScores(
            _key,
            start,
            end,
            Order.Descending);
        return HydrateEntries(entries, start + 1);
    }

    public int? GetRank(string userId)
    {
        var rank = _db.SortedSetRank(_key, userId, Order.Descending);
        return rank is null ? null : (int)rank + 1;
    }

    public IReadOnlyDictionary<string, string> GetUserMetadata(string userId)
    {
        var entries = _db.HashGetAll(MetadataKey(userId));
        return entries.ToDictionary(entry => (string)entry.Name!, entry => (string)entry.Value!);
    }

    public LeaderboardEntry? GetUserEntry(string userId)
    {
        var batch = _db.CreateBatch();
        var scoreTask = batch.SortedSetScoreAsync(_key, userId);
        var rankTask = batch.SortedSetRankAsync(_key, userId, Order.Descending);
        var metadataTask = batch.HashGetAllAsync(MetadataKey(userId));
        batch.Execute();
        Task.WaitAll(scoreTask, rankTask, metadataTask);

        if (scoreTask.Result is null || rankTask.Result is null)
        {
            return null;
        }

        return new LeaderboardEntry(
            Rank: (int)rankTask.Result + 1,
            UserId: userId,
            Score: scoreTask.Result.Value,
            Metadata: metadataTask.Result.ToDictionary(entry => (string)entry.Name!, entry => (string)entry.Value!),
            TrimmedUserIds: []);
    }

    public IReadOnlyList<LeaderboardEntry> ListAll()
    {
        var entries = _db.SortedSetRangeByRankWithScores(_key, 0, -1, Order.Descending);
        return HydrateEntries(entries, 1);
    }

    public long GetSize() => _db.SortedSetLength(_key);

    public bool DeleteUser(string userId)
    {
        var batch = _db.CreateBatch();
        var removeTask = batch.SortedSetRemoveAsync(_key, userId);
        var deleteMetadataTask = batch.KeyDeleteAsync(MetadataKey(userId));
        batch.Execute();
        Task.WaitAll(removeTask, deleteMetadataTask);
        return removeTask.Result;
    }

    public void Clear()
    {
        var userIds = _db.SortedSetRangeByRank(_key);
        var keys = new List<RedisKey> { _key };
        keys.AddRange(userIds.Select(userId => (RedisKey)MetadataKey(userId!)));
        _db.KeyDelete(keys.ToArray());
    }

    private string MetadataKey(string userId) => $"{_key}:user:{userId}";

    private static Dictionary<string, string> CoerceMetadata(IReadOnlyDictionary<string, string>? metadata) =>
        metadata is null ? [] : new Dictionary<string, string>(metadata);

    private IReadOnlyList<string> TrimToMaxEntries()
    {
        var overflow = GetSize() - _maxEntries;
        if (overflow <= 0)
        {
            return [];
        }

        var trimmedUserIds = _db.SortedSetRangeByRank(_key, 0, overflow - 1)
            .Select(value => (string)value!)
            .ToList();
        if (trimmedUserIds.Count == 0)
        {
            return [];
        }

        _db.SortedSetRemoveRangeByRank(_key, 0, overflow - 1);
        _db.KeyDelete(trimmedUserIds.Select(userId => (RedisKey)MetadataKey(userId)).ToArray());
        return trimmedUserIds;
    }

    private IReadOnlyList<LeaderboardEntry> HydrateEntries(SortedSetEntry[] entries, int startRank)
    {
        if (entries.Length == 0)
        {
            return [];
        }

        var batch = _db.CreateBatch();
        var metadataTasks = new List<Task<HashEntry[]>>();
        foreach (var entry in entries)
        {
            metadataTasks.Add(batch.HashGetAllAsync(MetadataKey(entry.Element!)));
        }

        batch.Execute();
        Task.WaitAll(metadataTasks.ToArray());

        return entries.Select((entry, index) => new LeaderboardEntry(
            Rank: startRank + index,
            UserId: entry.Element!,
            Score: entry.Score,
            Metadata: metadataTasks[index].Result.ToDictionary(
                hashEntry => (string)hashEntry.Name!,
                hashEntry => (string)hashEntry.Value!),
            TrimmedUserIds: [])).ToList();
    }

    private static int NormalizePositiveInt(int value, string fieldName) =>
        value < 1
            ? throw new ArgumentOutOfRangeException(fieldName, $"{fieldName} must be at least 1")
            : value;
}
